File: examples/cpa.rs
use anyhow::{Context, Result};
use gnuplot::{Figure, PlotOption::Caption};
use muscat::{
    distinguishers::cpa::CpaProcessor, leakage_model::aes::sbox, util::chipwhisperer_float_to_u16,
};
use ndarray::Array2;
use ndarray_npy::read_npy;
use std::{env, iter::zip, path::PathBuf};

fn leakage_model(plaintext_byte: usize, guess: usize) -> usize {
    sbox((plaintext_byte ^ guess) as u8) as usize
}

fn main() -> Result<()> {
    let traces_dir =
        PathBuf::from(env::var("TRACES_DIR").context("Missing TRACES_DIR environment variable")?);

    let traces: Array2<f64> =
        read_npy(traces_dir.join("traces.npy")).context("Failed to read traces.npy")?;
    let plaintexts: Array2<u8> =
        read_npy(traces_dir.join("plaintexts.npy")).context("Failed to read plaintexts.npy")?;
    assert_eq!(traces.shape()[0], plaintexts.shape()[0]);

    // Let's recover the first byte of the key
    let mut processor = CpaProcessor::new(traces.shape()[1], 256);
    for (trace, plaintext) in zip(traces.rows(), plaintexts.rows()) {
        processor.update(
            trace.mapv(chipwhisperer_float_to_u16).view(),
            plaintext[0],
            leakage_model,
        );
    }

    let cpa = processor.finalize(leakage_model);

    let best_guess = cpa.best_guess();
    println!("Best subkey guess: {best_guess:?}");

    // Let's plot correlation coefficients of the best guess
    let corr = cpa.corr();
    let corr_best_guess = corr.row(best_guess);

    let mut fg = Figure::new();
    fg.axes2d().lines(
        0..corr_best_guess.len(),
        corr_best_guess,
        &[Caption("Pearson correlation coefficient")],
    );
    fg.show()?;

    Ok(())
}

File: muscatpy/examples/cpa.py
import functools
import os
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

import muscatpy


@functools.cache
def hamming_weight(x: int) -> int:
    weight = 0
    while x != 0:
        if x & 1 == 1:
            weight += 1
        x = x >> 1
    return weight


TRACES_DIR = Path(os.environ["TRACES_DIR"])

traces = np.load(TRACES_DIR / "traces.npy")
plaintexts = np.load(TRACES_DIR / "plaintexts.npy")

assert traces.shape[0] == plaintexts.shape[0]

# We convert from np.float64 to np.float32 as f64 is not supported
traces = traces.astype(np.float32)
plaintexts = plaintexts.astype(np.uint64)

# Compute the CPA using the `compute_cpa` helper
cpa = muscatpy.distinguishers.compute_cpa(
    traces,
    plaintexts,
    256,
    0,
    lambda plaintext, guess: hamming_weight(
        muscatpy.leakage_model.aes.sbox(plaintext ^ guess)
    ),
    200,
)

# Or using the lower level `CpaProcessor` class
leakage_model = lambda plaintext, guess: hamming_weight(
    muscatpy.leakage_model.aes.sbox(plaintext ^ guess)
)

cpa_processor = muscatpy.distinguishers.CpaProcessor(traces.shape[1], 256, traces.dtype)
for i in range(0, traces.shape[0], 200):
    cpa_processor.batch_update(
        traces[i : i + 200], plaintexts[i : i + 200, 0], leakage_model
    )
cpa = cpa_processor.finalize(leakage_model)

best_guess = cpa.best_guess()
print("Best subkey guess:", best_guess)

corr = cpa.corr()[best_guess, :]
plt.plot(corr)
plt.title("Pearson correlation coefficient")
plt.show()
